tools/psa/build_tfm.py

```diff
@@ -0,0 +1,143 @@
+    Commands are passed as a list of tokens.
+    E.g. The command 'git remote -v' would be passed in as:
+     ['git', 'remote', '-v']
+
+    :param command: System command as a list of tokens
+    """
+    output = ''
+    logger.debug('[Exec] %s', ' '.join(command))
+    try:
+        fnull = open(os.devnull, 'w')
+        output = subprocess.check_output(command, stderr=fnull)
+    except subprocess.CalledProcessError as e:
+        logger.error("The command %s failed with return code: %s",
+                     (' '.join(command)), e.returncode)
+        sys.exit(1)
+    return output
+
+def detect_and_write_tfm_version(tfm_dir):
+    """
+    Identify the version of TF-M and write it to VERSION.txt
+    :param tfm_dir: The filesystem path where TF-M repo is cloned
+    """
+    cmd = ['git', '-C', tfm_dir, 'describe', '--tags',
+           '--abbrev=12', '--dirty', '--always']
+    tfm_version = run_cmd_and_return_output(cmd)
+    logger.info('TF-M version: %s', tfm_version)
+    if not isdir(VERSION_FILE_PATH):
+        os.makedirs(VERSION_FILE_PATH)
+    with open(join(VERSION_FILE_PATH, 'VERSION.txt'), 'w+') as f:
+        f.write(tfm_version)
+
+def check_repo_version(name, deps):
+    """
+    Compare the version of cloned and expected and exit if they dont match
+    :param name: Name of the git repository
+    :param deps: Dictionary containing dependency details
+    """
+    basedir = TF_M_BUILD_DIR
+    if name == 'trusted-firmware-m':
+        cmd = ['git', '-C', join(basedir, name),
+               'rev-parse', '--abbrev-ref', 'HEAD']
+    else:
+        cmd = ['git', '-C', join(basedir, name),
+               'describe', '--tags']
+    _out = run_cmd_and_return_output(cmd)
+    if _out.strip('\n') != deps.get(name)[1]:
+        logger.error('Conflict: cloned "%s" and expected "%s"',
+                     _out.strip('\n'), deps.get(name)[1])
+        logger.error('check and remove folder %s',
+                     join(basedir, name))
+        sys.exit(1)
+    else:
+        logger.info('%s: version check OK', name)
+
+def check_and_clone_repo(name, deps):
+    """
+    Test if the repositories are already cloned. If not clone them
+    :param name: Name of the git repository
+    :param deps: Dictionary containing dependency details
+    """
+    basedir = TF_M_BUILD_DIR
+    if not isdir(join(basedir, name)):
+        logger.info('Cloning %s repo', name)
+        cmd = ['git', '-C', basedir, 'clone', '-b',
+               deps.get(name)[1], deps.get(name)[0]]
+        _out = run_cmd_and_return_output(cmd)
+        logger.info('Cloned %s repo successfully', name)
+    else:
+        logger.info('%s repo exists, checking git version...', name)
+        check_repo_version(name, deps)
+
+def clone_tfm_repo():
+    """
+    Clone TF-M git repos and it's dependencies
+    """
+    check_and_clone_repo('trusted-firmware-m', dependencies)
+    check_and_clone_repo('mbedtls', dependencies)
```

Comment:
So, this puts tf-m and all dependencies side by side into a single directly. How does TF-M pick up the dependencies and use them?

Reply:
Yeah, that's the preferred way of building TF-M. The dependecies are cloned outside the TF-M source tree and the cmake scripts handle the inclusion of dependent modules.

Reply:
OK, so we can point TF-M's build system to any filesystem location for the dependencies.

Reply:
I need to check the cmake files but I think it is parent folder of tf-m is where the cmake will look for dependencies.